Add test to make sure we release the ICorProfilerCallback interfaces

davmason opened the pull request:
A recurring issue is that we add a new ICorProfilerCallback* interface and then forget to add the code to ~EEToProfInterfaceImpl to release it. See #41791 and #10756 for examples.

This test verifies that its destructor is called after detaching. As far as I can reason there isn't a good way to directly assert that all ICorProfilerCallback* interfaces are released, but in order to test new APIs we would have to change the base Profiler test class to inherit from that interface. So, unless we check in code without tests this will provide coverage.

davmason committed:
add test to make sure we release the ICorProfilerCallback interfaces

davmason committed:
fix casing for non-windows

davmason commented:
@noahfalk @sywhang @dotnet/dotnet-diag does anybody have time to review this?

sdmaclea commented on src/tests/profiler/unittest/releaseonshutdown.cs:
```diff
@@ -0,0 +1,66 @@
+
+            WasteTime();
+
+            return 100;
```
I don't see how this ever returns a failure.  Seems like `ReleaseOnShutdown::~ReleaseOnShutdown()` should set some globally observable state on a failure and change the result code.   It could also allow for an early out of the time based sleep if there the detach is complete.

davmason replied:
This doesn't ever return a failure. Profiler tests are different than normal tests. They relaunch themselves with a profiler attached and then monitor the stdout, in order to pass they must return 100 and have the output contain the phrase "PROFILER TEST PASSES" in stdout.

Most managed profilees will unconditionally return 100, the verification is usually done in the native profiler and it will print the pass message if everything succeeds. The only profilees that don't return 100 are ones that the profiler modifies via rejit or other mechanisms.

sdmaclea replied:
This code is in the relaunched profiler.  It could easily return a failure code, I understand that it doesn't need to...

It could still be relatively easy for the test to terminate when the destructor sets a process static variable to say the destructor had been run.  This would meet @josalem's concern

sdmaclea reviewed (changes requested):
+ Terminate the wait loop on test complete
+ Return result!=100 when test fails.

josalem commented on src/tests/profiler/unittest/releaseonshutdown.cs:
```diff
@@ -0,0 +1,66 @@
+            Console.WriteLine("Sleeping...");
+            for (int i = 0; i < 100; ++i)
+            {
+                Thread.Sleep(100);
```
Timeouts are difficult in CI.  Is there a way to sync with the profiler to know when you can stop waiting, e.g., poll for a file the profiler creates on detach?  1 second might not be long enough in all CI scenarios, e.g., gcstress, jitstress, etc., and I'd rather not artificially increase the length of the test.

sywhang replied:
I agree with John here - while adding the code to do this would be tedious, the time taken to diagnose a failure that happened because of such timeout is painful... And we learned this the hard way :-)

josalem reviewed (commented):
I'm happy to see a test to catch this scenario 👍

sywhang commented on src/tests/profiler/native/releaseonshutdown/releaseonshutdown.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "../profiler.h"
+
+#include <unordered_set>
```
nit: I don't think these headers ever used in this test?

davmason replied:
They aren't used, I think I used them in the first iteration of the test but it was before vacation and I can't remember why

davmason committed:
change sleep to waiting for a bool to be set by the profiler

davmason committed:
Add diagnostic output and disable warning that doesn't impact test

davmason committed:
Change misleading name ReleaseOnShutdown to ReleaseOnDetach

We don't actually release any ICorProfilerCallback* on shutdown, we only do on detach

sdmaclea commented on src/tests/profiler/native/releaseondetach/releaseondetach.cpp:
```diff
@@ -0,0 +1,120 @@
+{
+    if (_doneFlag != NULL)
+    {
+        *_doneFlag = true;
```
It would be better if:
+ this was set using an atomic operation to guarantee memory ordering on arm/arm64.
+ set after the `printf`s  completed.  maybe add a `fflush(stdout)` too

davmason replied:
What is the concern for the memory ordering on arm? I'm happy to change it if there's an issue, I just don't understand what the potential issue is.

Since the code that reads it is sitting in a spin loop, wouldn't the exact order of operations not matter? I.e. if it races with another thread it will just read it again on the next iteration?

sdmaclea replied:
It is just that done is set before the `printf` (so we are not quite done).  The spin loop could possibly terminate the test before the printing of the result completed.  I am not sure how quickly it could/would force the shutdown.  It is simply my defensive programming instinct.

davmason replied:
Oh, ok. I see, you're concerned about whether the success message gets printed to the console. I wasn't thinking about it from that angle.

As far as I know there isn't a cross platform InterlockedExchange equivalent. I marked it as volatile, added the fflush and moved it to the end of the method. I think that's all I can do without a lot of effort.

noahfalk replied:
> As far as I know there isn't a cross platform InterlockedExchange equivalent

If our compilers are supporting at least C++ 11 (I think they do?) then we should have access to std::atomic
https://docs.microsoft.com/en-us/cpp/standard-library/atomic-functions?view=vs-2019#atomic_compare_exchange_strong

davmason replied:
@noahfalk atomic works if you want an int/whatever built in type with atomic semantics but it doesn’t have a way to pass it an arbitrary pointer, does it? The bool * is coming from managed code. I could refactor and have an atomic bool in the profiler that gets passed to managed code, but I just don’t think it’s worth refactoring the test again.

sdmaclea reviewed (approved):
Nits

davmason committed:
remove unused headers and move flag set to after writing success message

noahfalk commented on src/tests/profiler/unittest/releaseondetach.cs:
```diff
@@ -0,0 +1,94 @@
+            bool profilerSetFlag = false;
+            for (int i = 0; i < 100; ++i)
+            {
+                Thread.Sleep(TimeSpan.FromSeconds(1));
```
I'd recommend polling with far lower latency, say 1ms 100,000 times. No single sleep makes a huge difference on its own, but historically one of the biggest areas of unnecessary delay in running a test pass were tests that spent a lot of time sleeping. Usually that in turn was slow polling intervals, waiting N sec for events that usually happen within a few ms.

noahfalk reviewed (approved):
LGTM : )

davmason committed:
change wait interval to one ms
